src/main/java/net/dv8tion/jda/core/JDABuilder.java

```diff
@@ -524,11 +526,41 @@ public JDA buildAsync() throws LoginException, IllegalArgumentException, RateLim
      */
     public JDA buildBlocking() throws LoginException, IllegalArgumentException, InterruptedException, RateLimitedException
     {
+        ReadyListener listener = new ReadyListener();
+        this.listeners.add(listener);
         JDA jda = buildAsync();
-        while(jda.getStatus() != Status.CONNECTED)
+        synchronized (listener)
         {
-            Thread.sleep(50);
+            while (!listener.isReady())
+            {
+                listener.wait();
+            }
```

Comment:
JDA internals should not be using event listeners as user implementations could hurt this system. The old loop was fine.